### PR TITLE
1.0.0-rc0

### PR DESCRIPTION
**Pull request by debel27**
Here is my proposal for the first release candidate of of _sse.js_ v1.0.0. If anything comes to mind, feel free to comment.

Please, consider this implementation as a working draft. Many iterations over this release candidate are still necessary.

### General

 - [x] specify and implement the _middleware_ architecture
 - [x] add support for the EventSource polyfill by exporting a dedicated middleware
 - [ ] add more integration tests, mainly to extensively cover the `send` method and its variants
 - [ ] add some performance test to measure library's efficiency under heavy workload
 - [ ] add `CONTRIBUTING.md` to describe guidelines on how to contribute. 

### Design and architecture

 - [ ] add `docs/ARCHITECTURE.md` to describe design decisions, implementation choices and tradeoffs
 - [ ] study if the ES6-class based API [is still relevant](https://github.com/joshburgess/not-awesome-es6-classes).
 - [x] study the case of [semaphores](https://en.wikipedia.org/wiki/Semaphore_(programming)) to guarantee the order of two successive `send` method calls. They may be needed due to the asynchronous nature of `_applyForTarget`. **Conclusion:** _no need for semaphores because order is already guaranteed by the underlying implementation of `send()` and `end()`_

### CODE REVIEW

**Comment by debel27** [comment]
**Regarding the monorepo architecture**

I have structured this project as a monorepo for the following reasons :

- if not for anything else, it allows to easily separate actual implementation from development code and separate dependencies from devDependencies. This vastly eases the procedure to publish npm packages that contains only what is necessary.
- the monorepo will regroup related projects under the same banner (such as "official" middlewares, etc), making them easier to tack, integrate and test alo the _sse_ project.

This design choice is indeed open to discussion.

File: packages/sse/README.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    const sseService = new SSEService()
+    http.createServer((request, response) => {
+      if(request.method === 'GET' && request.url === '/sse')
+        sseService.register(request, response);
+    }).listen(8080);
+    
+    sseService.on('connection', sseId => {
+      sseService.send('hello', sseId); // sends data to a single response
+      sseService.send({event:'user-connected'}); // broadcasts data to all responses
+    });
```
Comment (rexxars) [inline packages/sse/README.md]: Could we use the  \`\`\`js notation for these code samples in order to get proper syntax highlighting?

File: packages/sse/README.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        * [sseId.isConnectionActive()](#sseidisconnectionactive)
+        * [sseId.sseService](#sseidsseservice)
+     * [Event: 'clientClose'](#event-clientclose)
+     * [Event: 'connection'](#event-connection)
+     * [Event: 'error'](#event-error)
+     * [sseService.close([cb])](#sseserviceclosecb)
+     * [sseService.register(req, res)](#sseserviceregisterreq-res)
+     * [sseService.resetLastEventId([cb])](#sseserviceresetlasteventidcb)
+     * [sseService.send(opts[, target[, cb]])](#sseservicesendopts-target-cb)
+     * [sseService.unRegister([target[, cb]])](#sseserviceunregistertarget-cb)
```
Comment (rexxars) [inline packages/sse/README.md]: Looking at this now, I think it looks a bit weird with capitalized `R`. Do you have any strong preference here, or could we name it `unregister`?

File: packages/sse/lib/sse-service.js [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+  'Content-Type': 'text/event-stream',
+  Connection: 'keep-alive',
+  'Cache-Control': 'no-cache',
+  'X-Accel-Buffering': 'no'
+};
+
+class SSEService extends EventEmitter {
+  constructor() {
+    super();
+    internal(this).secureId = uuidv4();
```
Comment (rexxars) [inline packages/sse/lib/sse-service.js]: This is the first code I've seen using this weakmap approach, and seems a bit exotic to me, personally. I would personally just use an underscore prefix to annotate them as "private". If you feel strongly about keeping it as-is, couldn't these assignments be simplified slightly, at least? Not calling `internal()` for each assignment?

File: packages/sse/lib/sse-service.js [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+    // So `sseService.register` can be used as a standalone function (as an express middleware, for instance)
+    this.register = this.register.bind(this);
+  }
+
+  get SSEID() {
+    return internal(this).SSEID;
+  }
+
+  get nbActiveConnections() {
```
Comment (rexxars) [inline packages/sse/lib/sse-service.js]: How about `numActiveConnections` instead?

File: packages/sse/lib/sse-service.js [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+    if (validationError !== null) {
+      res.writeHead(validationError.status);
+      res.end(JSON.stringify({ error: validationError.message }));
+      return;
+    }
+
+    const { activeSSEConnections, SSEID, secureId } = internal(this);
+
+    // Not registering a connection that has already been registered
+    if (getSSEIdFromResponse(res, SSEID) === null) {
```
Comment (rexxars) [inline packages/sse/lib/sse-service.js]: Personal preference would be to reverse this and return early instead. I think it makes the code a little easier to read, since I don't have to keep reading to find out there is no "else" clause.

File: packages/sse/lib/sse-service.js [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+const uuidv4 = require('uuid/v4');
+const async = require('async');
+const assert = require('./utils/assert');
+const { maybeFn } = require('./utils/maybe');
+const { SSEIDClassFactory } = require('./sse-id');
+
+const internal = require('./utils/weak-map').getInternal();
+
+const HEARTBEAT_INTERVAL = 15;
+const HEARTBEAT_MESSAGE = 'heartbeat';
```
Comment (rexxars) [inline packages/sse/lib/sse-service.js]: I feel the default heartbeat should be fairly short - a lot of clients could easily cause quite a lot of traffic being pinged every 15 seconds.

**Review by rexxars — COMMENTED** [review]
Solid work! There's a few coding style differences from how I'm used to doing things which I've commented on, not sure how strongly you feel about this stuff.

Certain parts are also (in my opinion) a bit overengineered (factory classes etc, where a simple function probably would have been sufficient), but I'm not going to die on some hill on these points ;)

**Comment by debel27** [comment]
Hi rexxars, thank you for your feedback !

The usage of a `SSEIDClassFactory` class clearly is an overkill. I see no problem in replacing it with a simple function, which is definitely more Javascript-friendly. Thanks for pointing that out.

I get your feelings that some parts of the codebase may have been overengineered. This possibly comes from a design decision that aims to guarantee package's soundness, in the sense that it should not be possible to mess with its internal state.

As an example, the specific usage of WeakMaps you saw in the codebase (with the `internal` function) ensures that private properties truly are private. It is not possible to access them outside the `SSEService`. This may look confusing, but it provides a whole lot of insurance : nobody will create SSEIDs on their own, or do something that might compromise the package's integrity.

While these choices offer great deal of protection against over-permissive Javascript, they go against common practices and lack simplicity. I have decided to accept the tradeoff, as it gives me great confidence in the codebase. This probably comes from my mindset that anything that runs on the server must be rock-solid.

Yet I am definitely open to discussion. Do you think we should go back on this design decision ?

File: packages/sse/README.md (hunk 1 above)
Comment (debel27) [reply packages/sse/README.md]: Fully agreed.

File: packages/sse/README.md (hunk 2 above)
Comment (debel27) [reply packages/sse/README.md]: I definitely prefer `unregister`. The reason I went for `unRegister` was because of my IDE considering the former as a spelling error and I wanted to prevent the same trouble to the package's users. But what the heck, if I start over-engineering spelling this is definitely a sign that something is wrong !

TL;DR : I'll change to `unregister`

File: packages/sse/lib/sse-service.js (hunk 6 above)
Comment (debel27) [reply packages/sse/lib/sse-service.js]: The choice of `15` directly comes from [the specification](https://html.spec.whatwg.org/multipage/server-sent-events.html#authoring-notes).

I am a bit confused by your comment, as you seem to imply that I should shorten the interval, but at the same time doing so would bring even more traffic (which you apparently whishes to avoid). 

Or maybe you were simply suggesting that we sending all heartbeat at once might be problematic for performance ?

File: packages/sse/lib/sse-service.js (hunk 3 above)
Comment (debel27) [reply packages/sse/lib/sse-service.js]: I totally get you on this one. I will work something out.

I got the technique from [a great MDN article](https://developer.mozilla.org/en-US/docs/Archive/Add-ons/Add-on_SDK/Guides/Contributor_s_Guide/Private_Properties) regarding the topic of private object properties. Don't get fooled by the deprecation warning that actually concerns the development of Mozilla's add-ons. 

As mentioned one of my comments, I find private properties to offer a great deal of confidence, especially for server-side packages. This is my opinion on the matter of course, and I am open to discussion.

In any case, I certainly agree this looks quite verbose. I can work something out for making this more user-friendly.

File: packages/sse/lib/sse-service.js (hunk 4 above)
Comment (debel27) [reply packages/sse/lib/sse-service.js]: Deal.

File: packages/sse/lib/sse-service.js (hunk 5 above)
Comment (debel27) [reply packages/sse/lib/sse-service.js]: I am a strong advocate of this syntax also. I missed it here. I will definitely reverse and return instead.

File: packages/sse/lib/sse-service.js (hunk 6 above)
Comment (rexxars) [reply packages/sse/lib/sse-service.js]: Sorry, I actually meant the message length! As far as I know, we could potentially just send an empty comment, which would reduce each ping from 12 to 3 bytes.

**Comment by rexxars** [comment]
Nah, as long as it's an conscious choice with good reasons behind it, I have no problem with it.
It's internals, after all - if we get any feedback from other contributors on it, we can consider changing it, but let's keep it as-is for now.

File: packages/sse/lib/sse-service.js (hunk 6 above)
Comment (debel27) [reply packages/sse/lib/sse-service.js]: Fair enough. 

For a very large number of connections this is not negligible, you're right. I'm changing it.

**Comment by debel27** [comment]
Allright. I'll add some `CONTRIBUTING.md` at the root of the project to make those design decisions clear for future contributors.

File: packages/sse/lib/sse-service.js (hunk 3 above)
Comment (debel27) [reply packages/sse/lib/sse-service.js]: Hi rexxars,

after wrapping my head around the problem, it quickly becomes obvious that you just can't simplify this syntax. And by design, so.

The whole benefit of private properties comes from the closure. Private properties are stored in an internal object that first needs to be queried from the WeakMap before we can do any assignment.

If truly that bothers you, I can always do something like the following, but that's pretty much about it :
```javascript
const internals = internal(this);
internals.secureId = uuidv4();
```
I would therefore leave it as it is.

**Comment by debel27** [comment]
I have updated the list of TODOs in my first post. Once all of them are checked, I think we'll have a solid candidate for a merge with master.

I have made some improvements to the codebase to make it more approachable for contributors. A `docs/ARCHITECTURE.md` file is still necessary to present some design decisions.

**Comment by debel27** [comment]
I just added support for [middlewares](https://github.com/debel88/sse.js/tree/1.0.0#sseserviceusemiddlewares). Middlewares are the fundation that will allow package's extensibility by third-party plugins.

[Server requirements](https://github.com/Yaffle/EventSource#server-side-requirements) from the `event-source-poylfill` NPM package have been [implemented as a middleware](https://github.com/debel88/sse.js/blob/1.0.0/lib/middlewares/event-source-polyfill.js) that shall be exposed directly from the package. To keep the core minimal, I would not enable it by default.

An API based on lifecycle hooks seemed to be a fitting approach to me. Please, feel free to comment/review this design.
